api/v1alpha4/clusterclass_types.go

```diff
@@ -0,0 +1,106 @@
+*/
+
+package v1alpha4
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=clusterclasses,shortName=cc,scope=Namespaced,categories=cluster-api
+// +kubebuilder:storageversion
+
+// ClusterClass is a template which can be used to create managed topologies.
+type ClusterClass struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec ClusterClassSpec `json:"spec,omitempty"`
+}
+
+// ClusterClassSpec describes the desired state of the ClusterClass.
+type ClusterClassSpec struct {
+	// ControlPlane is a reference to a local struct that holds the details
+	// for provisioning the Control Plane for the Cluster.
+	ControlPlane LocalObjectTemplate `json:"controlPlane,omitempty"`
+
+	// Worker describes the worker nodes for the cluster.
+	// It is a collection of node types which can be used to create
+	// the worker nodes of the cluster.
+	// +optional
+	Worker WorkerClass `json:"worker,omitempty"`
+
+	// Infrastructure is a reference to a provider-specific template that holds
+	// the details for provisioning infrastructure specific cluster
+	// for the underlying provider.
+	// The underlying provider is responsible for the implementation
+	// of the template to an infrastructure cluster.
+	Infrastructure LocalObjectTemplate `json:"infrastructure,omitempty"`
+}
+
+// WorkerClass is a collection of deployment classes.
+type WorkerClass struct {
+	// Deployments is a list of machine deployment types that can be used to create
+	// a set of worker plane nodes.
+	Deployments []DeploymentClass `json:"deployments,omitempty"`
+}
+
+// DeploymentClass serves as a template to define a set of worker nodes of the cluster
+// provisioned using the `ClusterClass`.
+type DeploymentClass struct {
+	// Class denotes a type of worker node present in the cluster
+	Class string `json:"class,omitempty"`
+
+	// Template is a local struct containing a collection of templates for creation of
+	// MachineDeployment objects representing a set of worker nodes.
+	Template DeploymentTemplateClass `json:"template"`
+}
+
+// DeploymentTemplateClass defines how a MachineDeployment generated from a DeploymentClass
+// should look like.
+type DeploymentTemplateClass struct {
+	Metadata ObjectMeta `json:"metadata,omitempty"`
+
+	// Bootstrap contains the bootstrap template references to be used
+	// for the creation of worker Machines.
+	Bootstrap LocalObjectTemplate `json:"bootstrap,omitempty"`
+
+	// Infrastructure contains the infrastructure template references to be used
+	// for the creation of worker Machines.
+	Infrastructure LocalObjectTemplate `json:"infrastructure,omitempty"`
+}
+
+// LocalObjectTemplate defines a template for a topology Class.
+type LocalObjectTemplate struct {
+	// Ref is a required reference to a custom resource
+	// offered by a bootstrap provider.
+	Ref corev1.ObjectReference `json:"ref"`
```

Comment:
This object should have some validation that the `Ref` can only reference template objects, and `Kind` ends with `Template` suffix

Reply:
I'll move discussion about additional validation checks to a separated issue, this PR is already getting bigger than expected, so let's stick to the proposal...

Reply:
https://github.com/kubernetes-sigs/cluster-api/issues/4940